Use cookies for Auth, and general repo housekeeping

jamesdaniels opened the pull request:
Getting the NextJS codelab fixed up, this PR is intended to be shipped alongside [cl/744740598](https://critique.corp.google.com/cl/744740598)

* Update ignores to allow for easier local development
* Simply some code, namely using implied unsubscribe
* NextJS server actions are no longer experimental
* Drop service workers in favor of Cookies
    * fixes #306
    * fixes #295
    * fixes #294
    * fixes #298
    * fixes #297
    * fixes #299
    * fixes #308
* Added comments to explain `getAuthenticatedAppForUser`, fixes #274
* Fix restaurant details, fixes #292
* Use prettier, fixes #251 
* Have a default maxInstances of 2, should help for folks with policy limits like #302 #291

TODO:

- [x] Sync up narrative in [cl/744740598](https://critique.corp.google.com/cl/744740598)
- [x] Ensure that nextjs-start has all relevant changes
- [x] Do an unused import pass
- [x] Lint
- [x] Prettier

jamesdaniels committed:
Fixing the service worker config

jamesdaniels committed:
Simplify things

jamesdaniels committed:
Already covered by -log

jamesdaniels committed:
Dont crash service worker on network failures

jamesdaniels committed:
Cleanup

jamesdaniels committed:
Use magic url for sw sync

jamesdaniels committed:
More small cleanup

jamesdaniels committed:
Add comments

jamesdaniels committed:
Forgot miniify

jamesdaniels committed:
Eslint

jamesdaniels committed:
Use implied unsubscribe

jamesdaniels committed:
Port fixes from -end

jamesdaniels committed:
Sync up with -end

jamesdaniels committed:
more cleanup

jamesdaniels committed:
Use prettier

jamesdaniels committed:
Double quotes less disruptive

jamesdaniels committed:
es5 commas

jamesdaniels committed:
-start es5 comma

jamesdaniels committed:
Safety settings, runconfig

jamesdaniels committed:
Permit deployment on IDX out of box

jamesdaniels committed:
IDX template

jamesdaniels committed:
mv to $out

jamesdaniels committed:
Fix concurrently in idx

jamesdaniels committed:
IDX template

jamesdaniels committed:
cleanup those files

jamesdaniels committed:
Gitignore .env

jamesdaniels committed:
Dont need npx incantation, idx bug

jamesdaniels committed:
IDX setup

jamesdaniels committed:
Small fix

jamesdaniels committed:
asdf

jamesdaniels committed:
Trying no params

jamesdaniels committed:
Whoops

jamesdaniels committed:
Firebase setup

jamesdaniels committed:
bad sub

jamesdaniels committed:
firebase use on both

pashpashpash commented:
There's a problem with this if you log in with one user, then log in again as another user, it doesn't refresh and still uses the old user.

jamesdaniels commented:
@pashpashpash hmmmm, that's odd. What browser are you using? Are you seeing a problem with the service worker?

I did guard the router refresh behind the service worker ready & so I could ensure auth state was in sync with the new magic url method. 

I should do the same with login, only allow it if the service worker is registered and ready—since we rely on that.

pashpashpash commented:
Another issue I found is that server actions seem to drop the current user (currentUser=null), even though when SSR'ing a page that later calls the server action, the server says there is a currentUser.

pashpashpash commented:
Yeah, upon further inspection, it looks like server actions are sent as POST requests, which in your service worker setup:


```
self.addEventListener("fetch", (event) => {
  const { origin, pathname } = new URL(event.request.url);
  if (origin !== self.location.origin) return;

  // Use a magic URL to ensure that auth state is in sync between
  // the client and the service worker
  if (pathname.startsWith("/__/auth/wait/")) {
    const uid = pathname.split("/").at(-1);
    event.respondWith(waitForMatchingUid(uid));
    return;
  }

  if (pathname.startsWith("/_next/")) return;

  // Don't add headers to non-GET requests or those with an extension
  // This helps with CSS, images, fonts, JSON, etc.
  if (event.request.method !== "GET" || pathname.includes(".")) return;

  event.respondWith(fetchWithFirebaseHeaders(event.request));
});
```


Means that the firebase auth headers are not included in the request. So, naturally, the action tries to run:
`    const { firebaseServerApp, currentUser } = await getAuthenticatedAppForUser();`



And gets currentUser === null.

pashpashpash commented:
@jamesdaniels Okay I fixed this by allowing POST as well as GET requests in your service worker fetch event listener:



```

self.addEventListener("fetch", (event) => {
  const { origin, pathname } = new URL(event.request.url);
  if (origin !== self.location.origin) return;

  // Use a magic URL to ensure that auth state is in sync between
  // the client and the service worker
  if (pathname.startsWith("/__/auth/wait/")) {
    const uid = pathname.split("/").at(-1);
    event.respondWith(waitForMatchingUid(uid));
    return;
  }

  if (pathname.startsWith("/_next/")) return;

  // Don't add headers to non-GET requests or those with an extension
  // This helps with CSS, images, fonts, JSON, etc.
  if ((event.request.method === "GET" || event.request.method === "POST") && !pathname.includes(".")) {
    event.respondWith(fetchWithFirebaseHeaders(event.request));
  }
});


```



Please let me know if this is bad practice. But it seems to work fine. Idk

jamesdaniels commented:
@pashpashpash good catch thanks, i'll address that

jamesdaniels committed:
Early exit on head req too

jamesdaniels committed:
next-start

jamesdaniels committed:
Simplify, no router no state

pashpashpash commented:
Sorry to keep bugging you, but I'm still getting a weird bug where the auth state randomly tells me that the current user is null, when attempting a server action. 

Example scenario:

1. User clicks on a button on the client that fetches a server action
2. Server action calls `const { firebaseServerApp, currentUser } = await getAuthenticatedAppForUser();`, currentUser is authenticated and works fine.
3. User clicks on the button x amount of times again, and it works fine, user is authenticated
4. User clicks on the button, and all of a sudden, `currentUser` is null and the server action fails. 

The most frustrating part about this bug is it seemingly happens at random, it feels ephemeral. Often the button works fine, and subsequent button clicks with other server actions are authenticated. But then randomly later on it shows the user as unauthenticated again. No page refreshes or anything. Same page, just a different point in time.

payable-sebastien-kovacs commented:
> Sorry to keep bugging you, but I'm still getting a weird bug where the auth state randomly tells me that the current user is null, when attempting a server action.
> 
> Example scenario:
> 
> 1. User clicks on a button on the client that fetches a server action
> 2. Server action calls `const { firebaseServerApp, currentUser } = await getAuthenticatedAppForUser();`, currentUser is authenticated and works fine.
> 3. User clicks on the button x amount of times again, and it works fine, user is authenticated
> 4. User clicks on the button, and all of a sudden, `currentUser` is null and the server action fails.
> 5. User clicks on the button right after that failure (no page reload), and it works fine again.
> 
> The most frustrating part about this bug is it seemingly happens at random, it feels ephemeral. Often the button works fine, and subsequent button clicks with other server actions are authenticated. But then randomly later on it shows the user as unauthenticated again. No page refreshes or anything. Same page, just a different point in time.
> 
> EDIT: FYI: After some debugging, I was able to fix this by just retrying to get the authIdToken if the first time I tried it was null, i try 2 more times with a 0.25s sleep. Not sure why but it fixed the issue.
> 
> ```
> async function fetchWithFirebaseHeaders(request) {
>   let authIdToken = await getAuthIdToken();
>   if (!authIdToken) {
>     // sleep for 0.25s
>     await new Promise((resolve) => setTimeout(resolve, 250));
>     authIdToken = await getAuthIdToken();
>   }
>   if (!authIdToken) {
>     // sleep for 0.25s
>     await new Promise((resolve) => setTimeout(resolve, 250));
>     authIdToken = await getAuthIdToken();
>   }
>   if (authIdToken) {
>     const headers = new Headers(request.headers);
>     headers.append("Authorization", `Bearer ${authIdToken}`);
>     request = new Request(request, { headers });
>   }
>   return await fetch(request).catch((reason) => {
>     console.error(reason);
>     return new Response("Fail.", {
>       status: 500,
>       headers: { "Content-Type": "text/html" },
>     });
>   });
> }
> ```

Thank you both for looking at firebase with Next/RSC.

@pashpashpash Would you have a link to the whole service worker file that's working for you?

Also for curiosity - did you try setting / using cookies manually and to completely remove the service worker? I'm aware that this is probably going to prevent some FirebaseClient APIs to work correctly but I'm just trying to understand what other people have tried.

Thank you.

pashpashpash commented:
@payable-sebastien-kovacs 

here's the full file I have that's working well (as far as I've seen with gratuitous manual e2e testing):

```

// auth-service-worker.js

import { initializeApp } from "firebase/app";
import { getAuth, getIdToken, onAuthStateChanged } from "firebase/auth";

// Extract Firebase config from query string
const serializedFirebaseConfig = new URLSearchParams(self.location.search).get(
  "firebaseConfig"
);
if (!serializedFirebaseConfig) {
  throw new Error(
    "Firebase Config object not found in service worker query string."
  );
}

const firebaseConfig = JSON.parse(serializedFirebaseConfig);

const app = initializeApp(firebaseConfig);
const auth = getAuth(app);

self.addEventListener("install", () => {
  console.log("Service worker installed with Firebase config", firebaseConfig);
  self.skipWaiting();
});

self.addEventListener("activate", (event) => {
  event.waitUntil(self.clients.claim());
});

self.addEventListener("fetch", (event) => {
  const { origin, pathname } = new URL(event.request.url);
  if (origin !== self.location.origin) return;

  // Use a magic URL to ensure that auth state is in sync between
  // the client and the service worker
  if (pathname.startsWith("/__/auth/wait/")) {
    const uid = pathname.split("/").at(-1);
    event.respondWith(waitForMatchingUid(uid));
    return;
  }

  if (pathname.startsWith("/_next/")) return;

  // Don't add headers to non-GET requests or those with an extension
  // This helps with CSS, images, fonts, JSON, etc.
  if ((event.request.method === "GET" || event.request.method === "POST") && !pathname.includes(".")) {
    event.respondWith(fetchWithFirebaseHeaders(event.request));
  }
});

async function fetchWithFirebaseHeaders(request) {
  let authIdToken = await getAuthIdToken();
  if (!authIdToken) {
    // sleep for 0.25s
    await new Promise((resolve) => setTimeout(resolve, 250));
    authIdToken = await getAuthIdToken();
  }
  if (!authIdToken) {
    // sleep for 0.25s
    await new Promise((resolve) => setTimeout(resolve, 250));
    authIdToken = await getAuthIdToken();
  }
  if (authIdToken) {
    const headers = new Headers(request.headers);
    headers.append("Authorization", `Bearer ${authIdToken}`);
    request = new Request(request, { headers });
  }
  return await fetch(request).catch((reason) => {
    console.error(reason);
    return new Response("Fail.", {
      status: 500,
      headers: { "Content-Type": "text/html" },
    });
  });
}

async function waitForMatchingUid(_uid) {
  const uid = _uid === "undefined" ? undefined : _uid;
  await authStateReady();
  await new Promise((resolve) => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      if (user?.uid === uid) {
        unsubscribe();
        resolve();
      }
    });
  });
  return new Response(undefined, {
    status: 200,
    headers: { "Cache-Control": "no-store" },
  });
}

function authStateReady() {
  return new Promise((resolve) => {
    if (auth.currentUser !== undefined) {
      resolve();
    } else {
      const unsubscribe = onAuthStateChanged(auth, () => {
        unsubscribe();
        resolve();
      });
    }
  });
}

async function getAuthIdToken() {
  await authStateReady();
  if (!auth.currentUser) return null;
  return await getIdToken(auth.currentUser);
}

```

jamesdaniels committed:
Upgrade firebase, next, react; drop firebase-admin, idx stuff, and service worker in favor of Cookie auth

jhuleatt commented on nextjs-end/.gitignore:
```diff
@@ -1,4 +1,6 @@
 lib/firebase/config.js
+public/auth-service-worker.js
```
This isn't needed, since this file is deleted now, right?

jhuleatt replied:
done

jhuleatt commented on nextjs-end/.env:
We should get rid of this `.env` in favor of autoinit and `apphosting.yaml`

jhuleatt replied:
done

jhuleatt commented on nextjs-start/.env:
We should get rid of this `.env` in favor of autoinit and `apphosting.yaml`

jhuleatt replied:
done

jhuleatt commented on nextjs-start/.eslintrc:
```diff
@@ -0,0 +1,7 @@
+  "extends": ["eslint:recommended", "next", "prettier"],
+  "rules": {
+    "@next/next/no-img-element": "off",
+    "no-unused-vars": "off"
```
Is this because there are placeholder uninitialized variables in the starter code?

jamesdaniels replied:
yep

jhuleatt commented on nextjs-start/.gitignore:
```diff
@@ -1,4 +1,6 @@
 lib/firebase/config.js
+public/auth-service-worker.js
```
This isn't needed, since this file is deleted now, right?

jhuleatt replied:
done

jhuleatt commented on nextjs-end/firebase.json:
```diff
@@ -41,11 +41,7 @@
-    "ignore": [
-      "firebase.json",
-      "**/.*",
-      "**/node_modules/**"
-    ],
+    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
     "frameworksBackend": {
       "region": "us-central1"
     }
```
we should probably remove the whole hosting object to reduce confusion

jhuleatt replied:
There may be a web frameworks codelab floating around that uses this, so I'm leaving it in for now.

jhuleatt commented on nextjs-end/package-lock.json:
I thought this was gitignored?

jhuleatt commented on nextjs-end/package.json:
```diff
@@ -1,32 +1,39 @@
+  },
+  "devDependencies": {
+    "@next/eslint-plugin-next": "15.1.6",
+    "esbuild": "^0.20.2",
```
```suggestion
```

esbuild can be removed since we aren't building the service worker any more

jhuleatt replied:
done

jhuleatt commented on nextjs-start/package-lock.json:
I thought this was gitignored?

jhuleatt commented on nextjs-start/package.json:
```diff
@@ -1,32 +1,39 @@
+  },
+  "devDependencies": {
+    "@next/eslint-plugin-next": "15.1.6",
+    "esbuild": "^0.20.2",
```
```suggestion
```

we can remove esbuild now that we aren't building the service worker

jhuleatt replied:
done

jhuleatt commented on nextjs-end/src/app/page.js:
```diff
@@ -11,17 +11,21 @@ export const dynamic = "force-dynamic";
-		</main>
-	);
+export default async function Home(props) {
+  const searchParams = await props.searchParams;
```
whoa, TIL `searchParams` was changed to an async prop

jhuleatt commented on nextjs-end/src/lib/firebase/serverApp.js:
```diff
@@ -2,26 +2,28 @@
+// and Static Site Generation
 export async function getAuthenticatedAppForUser() {
-  const idToken = headers().get("Authorization")?.split("Bearer ")[1];
+  // Firebase App Hosting currently does not support cookies, so we transmit
+  // the client idToken via an Authorization header with Service Workers
```
```suggestion
```

jhuleatt replied:
done

jhuleatt commented on nextjs-end/src/lib/firebase/serverApp.js:
```diff
@@ -2,26 +2,28 @@
-  );
+  // Firebase Server App is a new feature in the JS SDK that allows you to
+  // instantiate the SDK with credentials retrieved from the client & has
+  // other afforances for use in server environments.
```
```suggestion
  // other affordances for use in server environments.
```

jhuleatt replied:
done

jhuleatt commented on nextjs-start/src/app/page.js:
need to do async searchparams here too

jhuleatt replied:
done

jhuleatt commented on nextjs-start/src/lib/firebase/clientApp.js:
```diff
@@ -1,12 +1,13 @@
-export const firebaseApp =
-  getApps().length === 0 ? initializeApp(firebaseConfig) : getApps()[0];
+
+export const firebaseApp = initializeApp(firebaseConfig);
```
TODO: autoinit

jhuleatt replied:
https://github.com/firebase/firebase-js-sdk/issues/8863#issuecomment-2751401913

jhuleatt replied:
done

jhuleatt commented on nextjs-end/src/lib/firebase/clientApp.js:
```diff
@@ -1,12 +1,13 @@
-export const firebaseApp =
-  getApps().length === 0 ? initializeApp(firebaseConfig) : getApps()[0];
+
+export const firebaseApp = initializeApp(firebaseConfig);
```
TODO: autoinit

jhuleatt replied:
https://github.com/firebase/firebase-js-sdk/issues/8863#issuecomment-2751401913

jhuleatt replied:
done

jhuleatt commented on nextjs-start/src/app/restaurant/[id]/page.jsx:
```diff
@@ -13,11 +16,14 @@ import { getFirestore } from "firebase/firestore";
+  const { firebaseServerApp } = await getAuthenticatedAppForUser();
+  const restaurant = await getRestaurantById(
+    getFirestore(firebaseServerApp),
+    params.id
```
need to do async params here too

jhuleatt replied:
done

jhuleatt commented on nextjs-end/src/components/RestaurantListings.jsx:
```diff
@@ -11,105 +11,98 @@ import { getRestaurantsSnapshot } from "@/src/lib/firebase/firestore.js";
+
+  useEffect(() => {
+    routerWithFilters(router, filters);
+  }, [router, filters]);
```
good catch

jhuleatt commented on nextjs-end/src/components/Reviews/ReviewSummary.jsx:
```diff
@@ -11,15 +15,23 @@ export async function GeminiSummary({ restaurantId }) {
   const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
-  const model = genAI.getGenerativeModel({ model: "gemini-pro"});
+  const model = genAI.getGenerativeModel({
+    model: "gemini-1.5-flash",
```
todo, we should update gemini version

jhuleatt replied:
switched to Genkit, updated model.

jhuleatt committed:
remove service worker from gitignores now that we're using cookie auth

jhuleatt committed:
use automatic initialization

https://firebase.google.com/docs/app-hosting/firebase-sdks#initialize-with-no-arguments

jhuleatt committed:
remove unused esbuild dep

this was previously used to generate the service worker

jhuleatt committed:
fix typo

jhuleatt committed:
async searchParams

jhuleatt committed:
async params

jhuleatt committed:
remove deprecated google ai SDK

use genkit instead

jhuleatt committed:
lint

jhuleatt committed:
remove outdated comment

jhuleatt committed:
update deps

jhuleatt committed:
re-run lint

jhuleatt commented on nextjs-start/src/components/Header.jsx:
```diff
@@ -1,76 +1,76 @@
+  signOut,
+  onIdTokenChanged,
 } from "@/src/lib/firebase/auth.js";
 import { addFakeRestaurantsAndReviews } from "@/src/lib/firebase/firestore.js";
```
```suggestion
import { addFakeRestaurantsAndReviews } from "@/src/lib/firebase/firestore.js";
import { setCookie, deleteCookie } from "cookies-next";
```

jhuleatt committed:
Add cookie imports to nextjs-start Header component

jhuleatt commented on nextjs-start/src/lib/firebase/serverApp.js:
```diff
@@ -3,11 +3,12 @@
 import "server-only";
 
 import { headers } from "next/headers";
```
```suggestion
import { cookies } from "next/headers";
```

jhuleatt committed:
Switch import to cookies in nextjs-start serverApp

lucasoares commented:
Hi @jamesdaniels or @jhuleatt. Is there any documentation or context on why the service worker was removed from this repository for Next.js applications?

I'm currently exploring best practices for integrating Firebase with a Next.js app. Firebase's own tutorials and examples don't use a service worker, but the Firehose site highlights several benefits of using one for authentication sessions: https://firebase.google.com/docs/auth/web/service-worker-sessions.

Given that, what is the currently recommended approach for integrating Firebase into production-ready Next.js applications?

bartcone commented:
@lucasoares +1 to that.
